### PR TITLE
Add Reverse Engineering (close #31)

### LINKED ISSUE #31: Rule 3, dev tools, and reverse engineering
https://blog.licensezero.com/2018/11/24/parity-5.0.0.html#rule-3 talks about licensing software I debug using a Parity-licensed debugger.

It’s pretty common to aim debuggers at source you don’t have (understanding what’s happening in closed-source code) or that isn’t necessarily source you can relicense (third-party published library code, maybe under that Artistic License with more restrictive terms you discuss).

I also read the rule as using the verbs in the habitual mode. If I add it into my compilation pipeline, I’m definitely developing my software with it. If I snag a linter once, aim it at my code as a trial, find it mostly sucks but fix one issue it identified, and never touch it again, I don’t think of it as a tool I developed with. Maybe that’s my mistaken conception, and one use is all it takes?

I also read “run with” as ambiguously “run alongside the software” (as part of a software system - like aiming one of those old disk compression utilities of the DOS age at a server’s disk volume) or “use to run (interpret and execute) the software” like if a Virtual Machine were Parity-licensed. I’m not clear on the intended reading - maybe both?

### PR DESCRIPTION
**Pull request by kemitchell**
@jeremy-w, finally got around to this! I would like to incorporate into the forthcoming Parity 7.

### CODE REVIEW

**Comment by zkat** [comment]
I like it. It does seem oddly specific to have this clause in the middle of it and I wonder if this is just one of many potential random exceptions to an otherwise straightforward license -- but I don't have a strong opinion against including this, and I think it does some work to emphasize that this license is partially being put out there to help create an open ecosystem?

**Comment by kemitchell** [comment]
@zkat I share your gut feeling about it. Actually, I feel the same, but moreso, about [Prototypes](https://github.com/licensezero/parity-public-license/blob/77c6bfa5f13f697a819ca65d5b50da85f012d8e9/LICENSE.md#prototypes). I don't want to write a big laundry list of exceptions, but I think these are very different exceptions than what we've seen in copyleft licenses before.

One the one hand, this looks like an even more special case, because it's fundamentally _about_ copyleft. Copyleft trips on itself when it stops open from taking on closed.

On the other hand, I could see toolmakers choosing Parity primarily for business reasons tripping up on it. If you choose Parity for your nifty new debugger because you want companies to pay you for a license, you may be unpleasantly surprised to see them using your debugger to clone a competitor's software for very competitive reasons.

Despite that footgun, I think it still belongs in the license. Backing up a step, Parity is designed to the "free for open source" license. I think reverse engineering to create open source alternatives is still "free for open source", even if the program actually debugged isn't.

File: LICENSE.md [hunk 1]
```diff
@@ -32,6 +32,10 @@ You don't have to [contribute](#contribute) any change, addition, or other softw
 
 3.  You don't develop, operate, or analyze software with it for anyone outside the team developing it.
 
+## Reverse Engineering
+
+You may use this software to operate and analyze software that you can't [contribute](#contribute) to develop alternatives that you can and do [contribute](#contribute).
```
Comment (jeremy-w) [inline LICENSE.md]: I hit a snag partway through reading this: I try to read “contribute to” as a phrasal verb, then realize right after that it’s “contribute [in order] to”.

Comment (kemitchell) [reply LICENSE.md]: I had "in order to" before. I'll add it back. Thanks!

**Comment by kemitchell** [comment]
@jeremy-w this license is short enough that I haven't reference every exception next to every rule. For example, Patent and Copyright don't say anything about Notices, Copyleft, or Defense. You have to read the (short) text as a whole.

File: LICENSE.md (hunk 1 above)
Comment (kemitchell) [reply LICENSE.md]: Addressed in 37fb535589417681ba4b866b5032796769432c0a

**Comment by kemitchell** [comment]
I'm going to merge this, since I'm already working a version with it anyway. But more feedback is _very_ welcome.

**Pull request merged**